public/static/app/perf-bb/PerfBBOutliersCtrl.test.js

```diff
@@ -39,211 +73,1443 @@ describe('PerfBBOutliersCtrlTest', () => {
+      const next_operation = {};
+      spyOn(vm.promises, 'isCurrent').and.returnValue(false);
+      spyOnProperty(vm.promises, 'next', 'get').and.returnValue(next_operation);
+      promise.finally = (success) => {
+        success();
+        return promise;
       };
+      vm.reload();
 
+      expect(vm.promises.isCurrent).toHaveBeenCalledWith(next_operation);
+      expect(vm.isLoading).toBe(true);
     });
 
-    it('should add type by default', () => {
-      const chain = scope.getAggChain({});
-      expect(chain.length).toEqual(2);
-      const filtering = chain.shift();
-      expect(filtering.length).toEqual(1);
-      expect(filtering[0].term).toEqual('=detected');
-      const sorting = chain.shift();
-      expect(sorting.$limit).toEqual(2000);
+  });
+
+  describe('lowConfidenceChanged', () => {
+
+    it('should call reload', () => {
+
+      spyOn(vm, 'reload').and.returnValue(undefined);
+      vm.lowConfidenceChanged();
+      expect(vm.reload).toHaveBeenCalled();
     });
 
-    it('should not add type when lowConfidence is set', () => {
-      const chain = scope.getAggChain({lowConfidence: true});
-      expect(chain.length).toEqual(1);
-      const sorting = chain.shift();
-      expect(sorting.$limit).toEqual(2000);
+  });
+
+  describe('modeChanged', () => {
+
+    it('should call reload', () => {
+
+      spyOn(vm, 'reload').and.returnValue(undefined);
+      vm.modeChanged();
+      expect(vm.reload).toHaveBeenCalled();
     });
 
-    it('should include filter if specified', () => {
-      const chain = scope.getAggChain({
-        filtering: {
-          filter1: 'value1',
-          filter2: 'value2',
+  });
+
+  describe('onRegisterApi', () => {
+
+    it('should register callbacks', () => {
+
+      const api = {
+        core:{
+          on:{
+            sortChanged: () => {},
+            filterChanged: () => {},
+            rowsRendered: () => {},
+          }
         }
+      };
+
+      spyOn(api.core.on, 'sortChanged').and.callThrough();
+      spyOn(api.core.on, 'filterChanged').and.callThrough();
+      spyOn(api.core.on, 'rowsRendered').and.callThrough();
+
+      const debounce = () => {};
+      const once = () => {};
+      const bind = () => {};
+      spyOn(_, 'debounce').and.returnValue(debounce);
+      spyOn(_, 'once').and.returnValue(once);
+      spyOn(_, 'bind').and.returnValue(bind);
+
+      vm.gridOptions.onRegisterApi(api);
+
+      expect(api.core.on.sortChanged).toHaveBeenCalledWith(scope, vm.state.onSortChanged);
+      expect(api.core.on.filterChanged).toHaveBeenCalledWith(scope, debounce);
+      expect(api.core.on.rowsRendered).toHaveBeenCalledWith(null, once);
+
+      expect(_.debounce).toHaveBeenCalledWith(vm.reload, 500);
+      expect(_.bind).toHaveBeenCalledWith(vm.state.onRowsRendered, vm.state);
+      expect(_.once).toHaveBeenCalledWith(bind, api);
+    });
+
+  });
+
+});
+describe('PerfBBOutliersFactoriesTest', () => {
```

Comment:
[nit] Can there be a blank line before this 'describe'?

Reply:
doh, I thought I'd gotten them all consistent.